src/com/google/enterprise/adaptor/database/ResponseGenerator.java

```diff
@@ -386,19 +395,80 @@ public void generateResponse(ResultSet rs, Response resp)
-      Blob blob = rs.getBlob(getContentColumnName());
-      InputStream in = blob.getBinaryStream();
-      OutputStream out = resp.getOutputStream();
-      com.google.enterprise.adaptor.IOHelper.copyStream(in, out);
-      in.close();
-      if (!(blob instanceof javax.sql.rowset.serial.SerialBlob)) {
-        // SerialBlob is adamant about not supporting free
-        try {
-          blob.free();
-        } catch (java.sql.SQLFeatureNotSupportedException | 
-            UnsupportedOperationException unsupported) {
-          // let JVM garbage collection deal with it
+
+      ResultSetMetaData rsMetaData = rs.getMetaData();
+      int columnType = Types.OTHER;
+      int numberOfColumns = rsMetaData.getColumnCount();
+      for (int i = 1; i < (numberOfColumns + 1); i++) {
+        String columnName = rsMetaData.getColumnLabel(i);
+        if (columnName.equals(getContentColumnName())) {
+          columnType = rsMetaData.getColumnType(i);
+          log.log(Level.FINEST, "Name: {0}, Type: {1}", new Object[] {columnName, columnType});
+        }
+      }
+
+      InputStream in = null;
+      Clob clob = null;
+      Blob blob = null;
+      switch (columnType) {
+        case Types.VARCHAR:
+          Reader rd = rs.getCharacterStream(getContentColumnName());
+          in = new ByteArrayInputStream(CharStreams.toString(rd).getBytes(Charsets.UTF_8));
+          break;
+        case Types.VARBINARY:
+          in = rs.getBinaryStream(getContentColumnName());
+          break;
+        case Types.CLOB:
+        case Types.LONGVARCHAR:
+          clob = rs.getClob(getContentColumnName());
+          Reader rd1 = clob.getCharacterStream();
+          in = new ByteArrayInputStream(CharStreams.toString(rd1).getBytes(Charsets.UTF_8));
+          break;
+        case Types.BLOB:
+        case Types.LONGVARBINARY:
+          blob = rs.getBlob(getContentColumnName());
+          in = blob.getBinaryStream();
+          break;
+        case Types.NCHAR:
+        case Types.NCLOB:
+        case Types.NVARCHAR:
+        case Types.LONGNVARCHAR:
+          Reader rd2 = rs.getNCharacterStream(getContentColumnName());
+          in = new ByteArrayInputStream(CharStreams.toString(rd2).getBytes(Charsets.UTF_8));
+          break;
+        case Types.SMALLINT:
+        case Types.TINYINT:
+        case Types.FLOAT:
+        case Types.DECIMAL:
+        case Types.NUMERIC:
+        case Types.DOUBLE:
+        case Types.INTEGER:
+        case Types.CHAR:
+          String content = rs.getString(getContentColumnName());
+          in = new ByteArrayInputStream(content.getBytes(Charsets.UTF_8));
+          break;
+        default:
+          break;
+      }
+
+      if (in != null) {
+        OutputStream out = resp.getOutputStream();
+        com.google.enterprise.adaptor.IOHelper.copyStream(in, out);
+        in.close();
+      }
+
+      try {
+        switch (columnType) {
+          case Types.CLOB:
+            clob.free();
```

Comment:
Just clean these up in-place above. It will make the code a little longer, but I think it will be more clear.